Add polynomial s shape impact function

chahank opened the pull request:
Changes proposed in this PR:
- 

This PR fixes #

### PR Author Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] Correct target branch selected (if unsure, select `develop`)
- [ ] Descriptive pull request title added
- [ ] Source branch up-to-date with target branch
- [ ] [Documentation](https://climada-python.readthedocs.io/en/latest/guide/Guide_PythonDos-n-Donts.html#2.--Commenting-&-Documenting) updated
- [ ] [Tests][testing] updated
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]
- [ ] [Changelog](https://github.com/CLIMADA-project/climada_python/blob/main/CHANGELOG.md) updated

### PR Reviewer Checklist

- [ ] Read the [Contribution Guide][contrib]
- [ ] [CLIMADA Reviewer Checklist](https://climada-python.readthedocs.io/en/latest/guide/Guide_Reviewer_Checklist.html) passed
- [ ] [Tests][testing] passing
- [ ] No new [linter issues][linter]

[contrib]: https://github.com/CLIMADA-project/climada_python/blob/main/CONTRIBUTING.md
[testing]: https://climada-python.readthedocs.io/en/latest/guide/Guide_Continuous_Integration_and_Testing.html
[linter]: https://climada-python.readthedocs.io/en/stable/guide/Guide_Continuous_Integration_and_Testing.html#3.C.--Static-Code-Analysis

peanutfun commented on climada/entity/impact_funcs/test/test_base.py:
```diff
@@ -60,6 +59,34 @@ def test_from_sigmoid(self):
+            inten, threshold=0.2, half_point=1, upper_limit=0.8, exponent=4,
+            haz_type='RF', impf_id=2, intensity_unit='m'
+            )
+        correct_mdd = np.array([0, 0.4, 0.76995746, 0.79470418, 0.79843158])
```
It's not clear why these are the correct values. You should check that the function is
* zero below the threshold
* zero at the threshold
* 0.5x the upper limit at the half_point *(this is covered)*
* (very) close to the upper limit somewhere 10x half_point
* behaving as expected when threshold < 0, half_point < threshold, upper_limit < 0, exponent < 0.

chahank replied:
In this case, I opted for a direct check of the values as one can easily recompute by hand what the correct values are. I think the checks you propose are less good as they do not directly test the mathematical formula. I can add some tests for the edge cases, although I think this is a clear overkill here.

peanutfun replied:
Please don't patronize the readers of your code ("can easily be computed"). Can't you find one or multiple sets of parameters where the true values can be calculated *without* a calculator?

> I can construct a function that fulfils all your tests but that does not represent the correct formula.

That is true for most tests. I agree that some tests have to be specific, but the corner cases have to be checked nonetheless.

peanutfun commented on climada/entity/impact_funcs/test/test_base.py:
```diff
@@ -60,6 +59,34 @@ def test_from_sigmoid(self):
+        correct_mdd = np.array([0, 0.4, 0.76995746, 0.79470418, 0.79843158])
+        self.assertTrue(np.array_equal(impf.paa, np.ones(5)))
+        np.testing.assert_array_almost_equal(impf.mdd, np.zeros(5))
+        self.assertTrue(np.array_equal(impf.intensity, np.arange(0, 5, 1)))
+        self.assertEqual(impf.haz_type, 'RF')
+        self.assertEqual(impf.id, 2)
+        self.assertEqual(impf.intensity_unit, 'm')
```
No need to test this again. If you really want to, please write a function instead of copy-pasting.

chahank replied:
I think it is good to test that the object remains consistent with the edge case.

chahank replied:
Done.

peanutfun reviewed (changes requested):
Thanks for the addition! I think the function will be very useful.

Please improve the documentation and the tests. One linter issue can also easily be fixed.

chahank commented on climada/entity/impact_funcs/base.py:
```diff
@@ -273,3 +273,70 @@ def set_sigmoid_impf(self, *args, **kwargs):
+        **kwargs):
+        """S-shape polynomial impact function hinging on four parameter.
+
+        f(I) = D(I)**exponent / (1 + D(I)**exponent) * upper_limit
+
+        D(I) = max[I - threshold, 0] / (half_point - threshold)
```
Thanks!

chahank replied:
The mathmode seems not to compile correctly. This was also already the case for the existing method `from_sigmoid_impf`. I will see if I can find the reason.

chahank commented on climada/entity/impact_funcs/base.py:
```diff
@@ -273,3 +273,70 @@ def set_sigmoid_impf(self, *args, **kwargs):
+
+        Parameters
+        ----------
+        intensity: tuple(float, float, float)
+            tuple of 3 intensity numbers along np.arange(min, max, step)
```
Ok, good to know! It does not matter in the case of the impact function in most cases, but good to improve, thanks!

chahank commented on climada/entity/impact_funcs/base.py:
```diff
@@ -273,3 +273,70 @@ def set_sigmoid_impf(self, *args, **kwargs):
+            of impacts.
+        half_point : float
+            Intensity at which 50% of maxixmum impact is expected.
+            Must be larger than thres.
```
Aaa yes, I had it first throw an error. Good catch.

chahank commented on climada/entity/impact_funcs/base.py:
```diff
@@ -273,3 +273,70 @@ def set_sigmoid_impf(self, *args, **kwargs):
+            Intensity at which 50% of maxixmum impact is expected.
+            Must be larger than thres.
+        upper_limit : float
+            Maximum impact value. Must be larger than 0.
```
Good point. But then it should maybe also be called `scale` parameter or something as it would become the lower limit ^^.

chahank commented on climada/entity/impact_funcs/base.py:
```diff
@@ -273,3 +273,70 @@ def set_sigmoid_impf(self, *args, **kwargs):
+        upper_limit : float
+            Maximum impact value. Must be larger than 0.
+        exponent: float
+            Exponent of the polynomial. Must be larger than 0.
```
Indeed, must is again too strong.

chahank replied:
I made the exponent to be positive. The function is senseless otherwise. Negative exponents now raises an error.

peanutfun replied:
But an exponent of exactly zero would work, right? The function would just be constant then (thinking about calibration: It might be better to have as few restrictions as possible to avoid setting strange parameter bounds)

chahank commented on climada/entity/impact_funcs/base.py:
```diff
@@ -273,3 +273,70 @@ def set_sigmoid_impf(self, *args, **kwargs):
+        if threshold > half_point:
+            mdd = np.zeros_like(intensity)
+        else:
+            D = (intensity - threshold) / (half_point - threshold)
```
For math formula I think single-letter variables are much clearer. Bad linter! But can do.

peanutfun commented on climada/entity/impact_funcs/base.py:
```diff
@@ -273,3 +273,74 @@ def set_sigmoid_impf(self, *args, **kwargs):
+        **kwargs):
+        """S-shape polynomial impact function hinging on four parameter.
+
+        .. math:: f(I) = \\frac{luk(I)**exponent}{1 + luk(I)**exponent}}\\cdot scale
+
+        .. math:: luk(I) = \\frac{max[I - threshold, 0]}{half_point - threshold}
```
Please write valid LaTeX. You don't have to excape characters here. You can use a single math directive for both formulae. Please use `\max` and either use `\mathrm{threshold}` or a single symbol, as in any math formula.

peanutfun commented on climada/entity/impact_funcs/base.py:
```diff
@@ -273,3 +272,83 @@ def set_sigmoid_impf(self, *args, **kwargs):
+
+        .. math::
+
+            f(I) = \frac{\textrm{luk}(I)^{\textrm{exponent}}}{1 + \textrm{luk}(I)^{\textrm{exponent}}}\cdot \textrm{scale} \\
+            \textrm{luk}(I) = \frac{\max[I - \textrm{threshold}, 0]}{\textrm{half_point} - \textrm{threshold}}
```
Fix line-too-long
```suggestion
            f(I) = \frac{\textrm{luk}(I)^{\textrm{exponent}}}{
                1 + \textrm{luk}(I)^{\textrm{exponent}}
            }
                \cdot \textrm{scale} \\
            \textrm{luk}(I) = \frac{\max[I - \textrm{threshold}, 0]}{
                \textrm{half_point} - \textrm{threshold}
            }
```

chahank replied:
Also for math mode? Ok, why not.

peanutfun commented on climada/entity/impact_funcs/base.py:
```diff
@@ -273,3 +272,83 @@ def set_sigmoid_impf(self, *args, **kwargs):
+        else:
+            luk = (inten - threshold) / (half_point - threshold)
+            luk[luk < 0] = 0
+            mdd = luk**exponent / (1 + luk**exponent) * scale
```
Clarify that `scale` is not in the denominator
```suggestion
            mdd = scale * luk**exponent / (1 + luk**exponent)
```

peanutfun commented on climada/entity/impact_funcs/test/test_base.py:
```diff
@@ -60,6 +59,50 @@ def test_from_sigmoid(self):
+        np.testing.assert_array_almost_equal(impf.mdd, np.zeros(5))
+        test_aux_vars(impf)
+
+        with self.assertRaises(ValueError):
```
Check actual error
```suggestion
        with self.assertRaisesRegex(ValueError, "Exponent value"):
```

chahank replied:
Thanks for the correction! Why should we priorize using `assertRaisesRegex` ?

peanutfun replied:
Because it tests the actual error message. `assertRaises` would succeed with *any* `ValueError` thrown, but you want to test for a specific one here.

Note that this is basically a shortcut for the idiom we were using so far:
```python
with self.assertRaises(Exception) as cm:
    do_stuff()
self.assertIn("Text", str(cm.exception))
```

chahank replied:
Makes total sense, thanks!

chahank commented:
Ok, I fixed all that you proposed, thanks! I also fixed one test and made sure that for `threshold = half_point` the method returns `mdd=0` (else there was a divide by 0).

peanutfun commented:
Great work! Will merge.
